AaC Documentation for the model root key (#659)
DevOps-MBSE/AaC: docs/source/project_documentation/user_guide/model.md

> # Model Definitions in AaC

Reviewer: Missed noticing this last night, but you need to add this to the User Guide index.rst file so that it is correctly exposed. 

docs/source/project_documentation/user_guide/index.rst

Author: Resolved in e3672be

The AaC `model` root key is used to represent the components of a system. These,
for example, can be the services in a system following a Microservice
Architecture (MSA) design.

In its most basic form, a `model` must have a `name` that distinguishes it from
other AaC definitions. At the same time, however, a `model` that defines nothing

> more than a `name` is of little use for anything more than representing external

Reviewer: At the same time, however, a `model` that defines nothing more than a `name` is only useful for representing external components or actors.

Author: I chose not to say "only useful" here because I don't generally like to make absolute statements (Could placeholder models be a valid use case for models with only a name? Could there be other uses that we haven't considered? etc). That said, I slightly modified your suggestion in commit e3672be. Does that work, @lizzcondrey?

Also, on a side note, you can use the "suggestion" button to make a suggestion that will show a diff of what you want changed which makes it a bit easier to see what to change or immediately apply the change.
<img width="441" alt="Screenshot 2023-08-29 at 11 00 11" src="https://github.com/DevOps-MBSE/AaC/assets/2924081/8643acf0-5791-42a4-942c-d8cfcb836067">

Reviewer: Yes, that rephrasing works. Better clarification than what I had. 

OMG thank you, haha. For this batch of edits I definitely wanted that ability but doing the review last night I didn't figure out how.

components or actors.

> It is recommended to include some amount of `behavior` definitions for each

Author: Resolved in e3672be

model as that allows users and the AaC package to understand the component as a
part of the system. Some questions we can answer for `model`s with defined

> `behavior`s are:

Reviewer 2: These behaviors can also be applied to the overall system model as opposed to just its components.

1. What does the component do?
1. What inputs does it expect?
1. What outputs can other components expect from this component?

## Defining Systems Through Model Composition

Whether your system is a monolith or a collection of microservices, the `model`
root key is used to represent how components fit together as part of the system
and what those components do as part of the system.

> To compose a system from individual `component`s, one uses the `components` key.

Reviewer 2: ```suggestion
To compose a system from individual `component`s, one specifies additional `models` in the system `model`'s `components` field.
```

> ```yaml
> model:
>   name: AlarmClock
>   description: A simple alarm clock
>   components:
>     - name: clock
>       type: Clock
>     - name: timer
>       type: ClockTimer
>     - name: alarm
>       type: ClockAlarm
>   behavior:
>     - ...
> ```

Reviewer 2: I've got the magic sauce for you.

```suggestion
```{eval-rst}
.. literalinclude:: ../../../../python/model/alarm_clock/alarm_clock.yaml
    :language: yaml
    :lines: 6-28
```
```

Reviewer: I was between pushing for leveraging the referencing ability for Sphinx or really just do it in a go with that task later this iteration. Late night brain went with the simple path of leave it for now because of time crunch. But you had the info so that's easy too! Thanks!

### Defining System Behavior In Models

Each model should define one, or more, `behavior`s that describe the component's
behaviors. These behaviors include a descrpition of `input`s, `output`s, and
`scenario`s that show how the model interacts with different components in the
system.

> Each `behavior` requires a `BehaviorType` which represents the method of

Reviewer 2: Possible edit -- I don't feel strongly about it.
```suggestion
Each `behavior` requires a `BehaviorType` which represents the mode of
```

Reviewer 2: It may also be beneficial to list some of the `BehaviorType` enums such as `TIMER` or `REQUEST_RESPONSE`

interaction for that behavior.

A `behavior` can also have any combination of `input`s and `output`s.

Additionally, the action of every behavior is represented by the `scenario`
structure. Each `scenario` describes necessary pre-conditions (`given`),
triggers (`when`), and post-conditions (`then`) for the behavior.

```yaml
model:
  name: AlarmClock
  description: A simple alarm clock
  components:
    - ...
  behavior:
    - name: setAlarm
      type: REQUEST_RESPONSE
      description: Set the alarm timer
      input:
        - name: targetTime
          type: Timestamp
      acceptance:

>         - scenario: setTimer

Reviewer 2: That's a good point -- I should have sought to provide a given step when creating the model. Could this be added a minor update story/task to the model then the documentation?

>         - scenario: setTimer

Author: I went ahead and added a basic `given` step. Let me know if there's anything else we need to resolve this. @lizzcondrey @Coffee2Bits

>         - scenario: setTimer

Reviewer: The added step makes sense to me, I don't think it will cause any issue adding it. Anything come to mind @Coffee2Bits?

> ```yaml
> model:
>   name: AlarmClock
>   description: A simple alarm clock
>   components:
>     - ...
>   behavior:
>     - name: setAlarm
>       type: REQUEST_RESPONSE
>       description: Set the alarm timer
>       input:
>         - name: targetTime
>           type: Timestamp
>       acceptance:
>         - scenario: setTimer
>           when:
>             - The user sets the alarm timer
>           then:
>             - The alarm is triggered when the clock reaches the specified time
> ```

Reviewer 2: More magic sauce.

```suggestion
```{eval-rst}
.. literalinclude:: ../../../../python/model/alarm_clock/alarm_clock.yaml
    :language: yaml
    :emphasize-lines: 18-23
    :lines: 6-28
```
```

Reviewer 2: Copy-pasting my resolution to the current build issues from Lizz's PR.

https://github.com/DevOps-MBSE/AaC/pull/660#pullrequestreview-1599581386

Author: I'll hold off on this change until Lizz's PR goes in.

Reviewer: Okay, that initial merge fixing this version conflict is in main now.

### Defining System State In Models

> TBD

Reviewer: Not a fan of TBD in documentation if it can be helped. Would rather something italicized to still make it distinct it's different/intentional.

_Pending an accurate model representation and sufficient implementation of `state` within models, this section will not be covered yet._

Reviewer 2: I agree with this.

Author: Resolved in e3672be
